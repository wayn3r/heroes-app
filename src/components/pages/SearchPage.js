import { useMemo } from 'react'
import { useNavigate } from 'react-router-dom'
import { useForm, useQueryParams } from '../../hooks'
import { getHeroesByName } from '../../selectors/getHeroesByName'
import HeroCard from '../atoms/HeroCard'

const SearchPage = () => {
    const navigate = useNavigate()
    const { q = '' } = useQueryParams()
    const [values, handleSearchChange] = useForm({
        search: q,
    })
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

    const handleSearchSubmit = e => {
        e.preventDefault()
        const { search } = values
        const searchQuery = search ? `?q=${search}` : ''
        navigate(searchQuery)
    }

    return (
        <div>
            <h1>Search Page</h1>
            <hr />

            <div className="row ">
                <div className="col-5">
                    <h4>
                        Search Form
                        <hr />
                        <form
                            onSubmit={handleSearchSubmit}
                            className="search-form"
                        >
                            <input
                                type="text"
                                placeholder="Find your hero..."
                                className="form-control col-8"
                                onChange={handleSearchChange}
                                autoComplete="off"
                                name="search"
                                value={values.search}
                            />
                            <button
                                type="submit"
                                className="btn btn-outline-primary"
                            >
                                Search
                            </button>
                        </form>
                    </h4>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {heroesFiltered.length > 0 ? (
                        heroesFiltered.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    ) : (
                        <div className="alert alert-info text-center">
                            No results...
                        </div>
                    )}
                </div>
            </div>
        </div>
    )
}

export default SearchPage
