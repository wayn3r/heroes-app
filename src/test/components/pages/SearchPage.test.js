import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import SearchPage from '../../../components/pages/SearchPage'
const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Pruebas en <SearchPage />', () => {
    const wrapper = mount(
        <MemoryRouter>
            <SearchPage />
        </MemoryRouter>
    )
    test('debe de mostrarse correctamente sin query params', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').text()).toBe('No results...')
    })

    test('debe de mostrar resultados y el input con el valor del queryString', () => {
        const hero = 'superman'
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=' + hero]}>
                <SearchPage />
            </MemoryRouter>
        )

        expect(wrapper.find('input').prop('value')).toBe(hero)
        expect(wrapper.find('.alert-info').exists()).toBe(false)
        expect(wrapper.find('HeroCard').exists()).toBe(true)
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de mostrar el mensaje de No results... si no encuentra el heroe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=no-real-hero']}>
                <SearchPage />
            </MemoryRouter>
        )
        expect(wrapper.find('.alert-info').text()).toBe('No results...')
        expect(wrapper.find('HeroCard').exists()).toBe(false)
    })

    test('debe de redireccionar al heroe seleccionado ', () => {
        const hero = 'batman'
        const preventDefault = jest.fn()
        wrapper
            .find('input')
            .simulate('change', { target: { name: 'search', value: hero } })
        wrapper.find('form').simulate('submit', { preventDefault })
        expect(preventDefault).toHaveBeenCalled()
        expect(mockNavigate).toHaveBeenCalledWith(`?q=${hero}`)
    })
})
