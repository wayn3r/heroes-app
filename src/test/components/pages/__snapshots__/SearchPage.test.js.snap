// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrar resultados y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Page
  </h1>
  <hr />
  <div
    className="row "
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
        <hr />
        <form
          className="search-form"
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            className="form-control col-8"
            name="search"
            onChange={[Function]}
            placeholder="Find your hero..."
            type="text"
            value="superman"
          />
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Search
          </button>
        </form>
      </h4>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card"
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Superman"
              className="card-img"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <div
                className="card-title"
              >
                Superman
              </div>
              <p
                className="card-text"
              >
                Kal-El
              </p>
              <p
                className="card-text"
              >
                Action Comics #1
              </p>
              <a
                href="/hero/dc-superman"
                onClick={[Function]}
              >
                MÃ¡s detalles...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente sin query params 1`] = `
<div>
  <h1>
    Search Page
  </h1>
  <hr />
  <div
    className="row "
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
        <hr />
        <form
          className="search-form"
          onSubmit={[Function]}
        >
          <input
            autoComplete="off"
            className="form-control col-8"
            name="search"
            onChange={[Function]}
            placeholder="Find your hero..."
            type="text"
            value=""
          />
          <button
            className="btn btn-outline-primary"
            type="submit"
          >
            Search
          </button>
        </form>
      </h4>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info text-center"
      >
        No results...
      </div>
    </div>
  </div>
</div>
`;
